aptos-move/framework/aptos-framework/sources/code.move

```diff
@@ -0,0 +1,183 @@
+        /// Source map, in internal encoding
+        source_map: vector<u8>,
+        /// Source text.
+        source: String,
+    }
+
+    /// A package dependency. This consists of an address and the name of a package
+    /// at this address.
+    struct PackageDep has store, copy, drop {
+        addr: address,
+        name: String,
+    }
+
+    /// An address alias.
+    struct AddressAlias has store, copy, drop {
+        alias: String,
+        addr: address
+    }
+
+    /// Describes an upgrade policy
+    struct UpgradePolicy has store, copy, drop {
+        policy: u8
+    }
+
+    /// A package is attempted to publish with module names clashing with modules published by other packages on this
+    /// address.
+    const EMODULE_NAME_CLASH: u64 = 0x1;
+
+    /// A package is attempted to upgrade which is marked as immutable.
+    const EUPGRADE_IMMUTABLE: u64 = 0x2;
+
+    /// A package is attempted to upgrade with a weaker policy than previously.
+    const EUPGRADE_WEAKER_POLICY: u64 = 0x3;
+
+    /// Whether unconditional code upgrade with no compatibility check is allowed. This
+    /// publication mode should only be used for modules which aren't shared with user others.
+    /// The developer is responsible for not breaking memory layout of any resources he already
+    /// stored on chain.
+    public fun upgrade_policy_no_compat(): UpgradePolicy {
+        UpgradePolicy{policy: 0}
+    }
+
+    /// Whether a compatibility check should be performed for upgrades. The check only passes if
+    /// a new module has (a) the same public functions (b) for existing resources, no layout change.
+    public fun upgrade_policy_compat(): UpgradePolicy {
+        UpgradePolicy{policy: 1}
+    }
+
+    /// Whether the modules in the package are immutable and cannot be upgraded.
+    public fun upgrade_policy_immutable(): UpgradePolicy {
+        UpgradePolicy{policy: 2}
+    }
+
+    /// Whether the upgrade policy can be changed. In general, the policy can be only
+    /// strengthened but not weakened.
+    public fun can_change_upgrade_policy_to(from: UpgradePolicy, to: UpgradePolicy): bool {
+        from.policy <= to.policy
+    }
+
+    /// Publishes a package at the given signer's address. The caller must provide package metadata describing the
+    /// package.
+    public fun publish_package(owner: &signer, pack: PackageMetadata, code: vector<vector<u8>>) acquires PackageRegistry {
+        let addr = signer::address_of(owner);
+        if (!exists<PackageRegistry>(addr)) {
+            move_to(owner, PackageRegistry{packages: vector::empty()})
+        };
+
+        // Check package
+        let module_names = get_module_names(&pack);
+        let packages = &mut borrow_global_mut<PackageRegistry>(addr).packages;
+        let len = vector::length(packages);
+        let index = len;
+        let i = 0;
+        while (i < len) {
+            let old = vector::borrow(packages, i);
+            if (old.name == pack.name) {
```

Comment:
Is it possible to publish a package with a different name that still contains a module with the same name as an existing module (effectively an upgrade)?

Reply:
No thats not possible. Because the package defines the upgrade policy, we cannot allow this. We may however introduce an independent transaction to rename a package to achieve this effect.

Reply:
I don't recall a validation to make sure this case doesn't happen in the vm's publish logic? Is it there somewhere but I missed it?

Reply:
It is not in the VM's logic (which does not know about packages) but in our own, as added by this PR.

Reply:
I see check_coexistence does this.